File: Structural-Patterns/Flyweight/Java/Factory/PlaceTemplateFactory.java
package Factory;

import java.util.HashMap;
import java.util.Map;

import MapPlaces.PlaceCategory;
import MapPlaces.PlaceTemplate;

public class PlaceTemplateFactory{
    private final Map<PlaceCategory, PlaceTemplate> templates = new HashMap<>();

    public PlaceTemplate getTemplate(PlaceCategory placeCategory){
        if(!templates.containsKey(placeCategory)){
            templates.put(placeCategory, new PlaceTemplate(placeCategory));
        }
        return templates.get(placeCategory);
    }

    public int getNumberOfTemplates(){
        return templates.size();
    }
}

File: Structural-Patterns/Flyweight/C#/Flyweight/Factory/PlaceTemplateFactory.cs
public class PlaceTemplateFactory{
    private readonly IDictionary<PlaceCategory, PlaceTemplate> templates = new Dictionary<PlaceCategory, PlaceTemplate>();

    public PlaceTemplate getTemplate(PlaceCategory placeCategory){
        if(!templates.ContainsKey(placeCategory)){
            templates.Add(placeCategory, new PlaceTemplate(placeCategory));
        }
        return templates[placeCategory];
    }

    public int getNumberOfTemplates(){
        return templates.Count;
    }
}